File: medium/82.cpp
struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(nullptr) {}
};

class Solution {
public:
    ListNode* deleteDuplicates(ListNode* head) {
      ListNode* p0 = head;
      ListNode** lp = &head;
      while (p0) {
        int val = p0->val;
        ListNode* p1 = p0->next;
        if (!p1 || p1->val != val) {
          *lp = p0;
          lp = &(p0->next);
          p0 = p1;
        } else {
          do {
            p1 = p1->next;
          } while (p1 && p1->val == val);
          p0->next = p1;
          p0 = p1;
        }
      }
      *lp = nullptr;
      return head;
    }
};

File: medium/82.go
func deleteDuplicates(head *ListNode) *ListNode {
  for {
    if head == nil {
      return nil
    }
    if head.Next == nil {
      return head
    }
    hv := head.Val
    if hv != head.Next.Val {
      break
    }
    next := head.Next.Next
    for ; next != nil && next.Val == hv; next = next.Next {
    }
    head = next
  }
  ret := head
  next := head.Next
  for next != nil {
    if next.Next == nil {
      break
    }
    nv := next.Val
    next1 := next.Next
    flag := true
    for ; next1 != nil && next1.Val == nv; next1 = next1.Next {
      flag = false
    }
    if flag {
      head.Next = next
      head = next
      next = next.Next
    } else {
      next = next1
    }
  }
  head.Next = next
  return ret
}
